### ISSUE TITLE
Multiple joins to same table (for query composition) results in joining the same table multiple times

### ISSUE THREAD

**Issue by jeffutter** [elixir-ecto/ecto#2361]
I'm creating a ticket for an issue that was previously mentioned in another issue #1284 but wasn't fully resolved there.

I am running into problems dynamically composing queries from separate functions that may have similar joins in them. This is the result of JSON API 'filters'. There is a function for each filter and since the filters are added dynamically there is no guarantee whether or not one would be executed before another to set up `joins` and bindings. The only option, in this case, is to include the join in each function. This results in having conditions scoped to one or the other join but not both as one may expect.

See below for a (hopefully) clearer example.

One solution for this was mentioned in #1284 as 'named joins'. I'm opening a new issue for this since #1284 is closed so that this doesn't lose visibility.

### Environment

* Elixir version (elixir -v):
Elixir 1.5.3

* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.):
PostgreSQL 9.4

* Ecto version (mix deps):

2.1.6

* Database adapter and version (mix deps):

postgrex 0.13.3

* Operating system:

MacOS

### Setup

```elixir

def query(credential_type, credential_status) do
  # This is a simplified example showing a composite filter, however, a query could be generated with either of these conditions or none of these conditions.
  Ecto.Query.from(p in Profile)
  |> required_credential_type(credential_type)
  |> required_credential_status(credential_status)
end

def required_credential_type(query, credential_type) do
  Ecto.Query.from(p in query,
                               join: c in assoc(p, :credential),
                               on: c.profile_id == p.id,
                               where: c.type == ^credential_type)
end

def required_credential_status(query, credential_status) do
  Ecto.Query.from(p in query,
                               join: c in assoc(p, :credential),
                               on: c.profile_id == p.id,
                               where: c.status == ^credential_status)
end
```

q = Ecto.Query.from(i in q, join: c in "credential", on: c.interpreter_profile_id == i.id, where: c.state == "active"); q = Ecto.Query.from(i in q, join: c in "credential", on: c.interpreter_profile_id == i.id, where: not is_nil(c.name)); Ecto.Adapters.SQL.to_sql(:all, Repo, q)

### Current behavior

```elixir
iex> query("my_credential", "approved")

#Ecto.Query<from p in Profile, join: c0 in "credential",
 on: c0.profile_id == i.id, join: c1 in "credential",
 on: c1.profile_id == i.id, where: c0.type == "my_credential",
 where: c1.state == "approved">
```

I believe this would find profiles that have either approved credentials _or_ credentials of the "my_credential" type.

### Expected behavior

```elixir
iex> query("my_credential", "approved")

#Ecto.Query<from p in Profile, join: c0 in "credential",
 on: c0.profile_id == i.id,
 where: c0.type == "my_credential",
 where: c0.state == "approved">
```

I would expect to get a query that only returns credentials that are _both_ approved and of the "my_credential" type.

I see the need for the flexibility of both approaches however it seems that the combining approach would be more common than the ability to join twice to the same table and (as far as I can tell) it is not supported.

If anyone has any ideas around this or workarounds it would be greatly appreciated.

**Comment by CarlosCota** [comment]
Hoping for the best

**Comment by michalmuskala** [comment]
This is expected behaviour and not something we're planning to change.

The suggested solution is to first collect all the necessary joins, perform them and later when adding conditions refer to the joined table through multiple bindings.

**Comment by jeffutter** [comment]
I believe on my project I can enumerate all of the possible joins up front and use the whole list of bindings on every query. However, on some projects, I imagine the number of potential joins would cause performance issues, joining on tables that aren't used in the query? Perhaps the Postgres query optimizer will ignore them?

If one wanted to dynamically add the joins, one would also have to dynamically create the binding list, correct? Is the best way to do that using `Code.eval_quoted` at runtime as in [ex_sieve](https://github.com/valyukov/ex_sieve/blob/master/lib/ex_sieve/builder/where.ex)? If so this seems like a fairly non-trivial bit of coding. I can't imagine users new to elixir/ecto being able to accomplish this.

**Comment by josevalim** [comment]
You don't need to always use the whole list of bindings. As long as you are doing a relationship between the source in `from` and the latest join, you can comfortably build the query using `...`, such as `where([from, ..., last_join], from.foo == last_join.bar)`.

It only gets complex when you need to do relationships between joins but that doesn't seem to be the case with `ex_sieve`.

**Comment by jeffutter** [comment]
@josevalim Thank you, I appreciate the suggestion.

Perhaps `ex_sieve` was a bad example. I only referenced it's strategy since I found it in an [elixir form](https://elixirforum.com/t/build-dynamic-bindings-for-ecto-query-order-by/1638) post that seemed related.

For sake of discussion, I believe the following example better demonstrates the type of issue I am running into:

```elixir
def query(options) do
  credential_status = Keyword.get(options, :credential_status)
  credential_type = Keyword.get(options, :credential_type)
  account_status = Keyword.get(options, :account_status)
  fund_amount = Keyword.get(options, :fund_amount)

  Ecto.Query.from(p in Profile)
  |> maybe_join_credential(credential_status || credential_type)
  |> maybe_join_account(account)
  |> maybe_join_fund(fund_amount)
  |> maybe_require_credential_status(credential_status)
  |> maybe_require_credential_type(credential_type)
  |> maybe_require_account_status(account_status)
  |> maybe_require_fund_amount(fund_amount)
end

defp maybe_join_credential(query, nil), do: query
defp maybe_join_credential(query, _) do
  Ecto.Query.from(p in query,
                  join: c in assoc(p, :credential),
                  on: c.profile_id == p.id)
end

defp maybe_join_account(query, nil), do: query
defp maybe_join_account(query, _) do
  Ecto.Query.from(p in query,
                  join: c in assoc(p, :account),
                  on: c.profile_id == p.id)
end

defp maybe_join_fund(query, nil), do: query
defp maybe_join_fund(query, _) do
  Ecto.Query.from(p in query,
                  join: c in assoc(p, :fund),
                  on: c.profile_id == p.id)
end

defp maybe_require_credential_status(query, nil), do: query
defp maybe_require_credential_status(query, credential_status) do
  Ecto.Query.where(query, [p, c], c.status == ^credential_status)
end

defp maybe_require_credential_type(query, nil), do: query
defp maybe_require_credential_type(query, credential_type) do
  Ecto.Query.where(query, [p, c], c.type == ^credential_type)
end

defp maybe_require_account_status(query, nil), do: query
defp maybe_require_account_status(query, account_status) do
  Ecto.Query.where(query, [p, c, a], a.status == ^account_status)
end

defp maybe_require_fund_amount(query, nil), do: query
defp maybe_require_fund_amount(query, fund_amount) do
  Ecto.Query.where(query, [p, c, a, f], f.amount >= ^fund_amount)
end
```

This works great for the following command:

```elixir
query([credential_status: "approved", credential_type: "credential_a", account_status: "active"])
```

But falls down for this:

```elixir
query([account_status: "active"])
```

Since there will be no `c` binding in the query in `maybe_require_account_status`. We also can't use `...` here since the `a` binding _could_ be the last one, but `f` may also be the last binding.

All-in-all I have a solution for my particular issue right now so there is no urgency to this issue for me, but it does seem like a possible scenario that Ecto doesn't elegantly support.

**Comment by michalmuskala** [comment]
In case like this, I'd do something like:

```
def query(options) do
  Profile
  |> credential_clauses(options)
  |> account_clauses(options)
  |> fund_amount_clauses(options
end
```

And each of the functions manages it's joins and adds needed conditions.

**Comment by OvermindDL1** [comment]
https://elixirforum.com/t/ecto-query-patterns-named-joins/8474

This use case is something that would be fixed by named joins as I keep hitting myself.  I'd love to gather the joins I need, then join them via names (no positions, you can see the horrors of workarounds I do in the above link), then do what I need.  SQL uses named joins, Ecto should too; positional joins are horrible to work with on many conditionally joined SQL statements.

**Comment by michalmuskala** [comment]
The thing is, that there's no way to support named joins without either completely breaking user applications (which means people will just stay on ecto 2.x forever, just like python) or making queries significantly slower.

If you have proposals for how to actually implement this, I'm open to hearing them - I think this is something great for the ecto mailing list.

**Comment by OvermindDL1** [comment]
> The thing is, there's no way I can think of to support named joins without either completely breaking user applications (which means people will just stay on ecto 2.x forever, just like python) or making queries significantly slower.

In essence the query structure would need to keep a name with the joined tables, but other than that I don't think that any other structure would really need to be changed.  Just have the table names, like at `select(query, [a, c, e: e, f: f], %{...})` would use positional lookups for `a` and `c` as it does now, and use names to lookup the join links for `e` and `f` (binding them to `e` and `f` respectively for this example, though could rename that for brevity if the name is long for example, like `some_big_join_name: s`).  When a join is done then the name would be specified as given in the `in`, perhaps defaulting to what is explicitly given (just 'use most recent name' or something) or defaulting to no-name unless specified, and could specify it explicitly via something like `join(query, ..., [the_name: c] in ..., ...)` so as to match the keyword list in lookups, or could do `join(query, ..., %{the_name: c} in ..., ...)` for explicit naming, or a 2-arg tuple, or whatever (I don't care much about what syntax is chosen for the macro, as long as the feature exists and is useful, though a keyword list would be the shortest I'd think).

> If you have proposals for how to actually implement this, I'm open to hearing them - I think this is something great for the ecto mailing list.

That involves opening an email client, which I don't actually have open very often, plus lack of time for a full proper example (of which if I had the time I'd just make the PR straight out, maybe later this year unless someone beats me to it).

And of course in the elixir forum thread I linked, no one responded with any other ideas for handling the same join patterns, if you have one then I'd love to know about it, my combinatorial explosion of cases is quite a bit painful on compile time...  ^.^;

**Comment by mindreader** [comment]
The elixir mailing list appears to be a google group, so just join it on the web and post.  It doesn't seem super active, but if that is what the developers use, then it is well worth starting a discussion there.

If it is a hard problem due to its current architecture, I understand.  I would absolutely be willing to put in some free time if I knew there would be a desire for it.

But in the scheme of things I'm not sure why I would use what amounts to a very large dsl for querying when it's not composable.  You might as well just use prepared sql at that point.

**Comment by josevalim** [comment]
This is implemented in master, see #2389.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by mindreader** [comment]
Awesome.  I very much look forward to trying it out.  It looks like you are going in the direction I was hoping for.
